app/controllers/application_controller.rb

```diff
@@ -97,11 +97,36 @@ def mnemosyne_trace
     end
   end
 
-  def switch_locale(&) # rubocop:disable Metrics/AbcSize
-    session[:locale] = sanitize_locale(params[:custom_locale] || params[:locale] || session[:locale])
-    locale = session[:locale] || http_accept_language.compatible_language_from(I18n.available_locales) || I18n.default_locale
+  def sanitized_locale_param
+    sanitize_locale(params[:locale])
+  end
+
+  def sanitized_session_locale
+    sanitize_locale(session[:locale])
+  end
+
+  def users_preferred_locale
+    return if current_user.nil?
+
+    sanitize_locale(current_user.preferred_locale)
+  end
+
+  def choose_locale
+    (sanitized_locale_param ||
+      sanitized_session_locale ||
+      users_preferred_locale ||
+      http_accept_language.compatible_language_from(I18n.available_locales) ||
+      I18n.default_locale).to_sym
+  end
```

Comment:
Since I added a not-NULL constraint to the preferred_locale-column we dont really need the browser locale and the default locale here. I wanted that constraint because I thought it would be nice to always have something valid in there. On the other hand we then cannot know whether the preferred locale of the user is actually :en or if the user just has not chosen his/her locale yet. I think the only case where this is not ideal is when a user has :de set as their preferred locale in their browser settings and the user first signs up. In this case we would take the default locale :en from the database and think this was the preferred locale of the user. However as soon as the user selects :de as the locale everything works again.

Reply:
Using a not-NULL constraint is fine for me, makes sense. The other behavior, with a fallback to English on user sign up, however, is a little unexpected. While I thought we could explicitly store the "current" locale on sign up, that's not easily possible for SAML sign ins. This case might need dedicated handling in the omniauth_controller to work similar. The latter further requires that we also receive the user's current cookie on the omniauth callback, which I am currently not 100% sure about (since a user's request is originating from a third-party domain). 

In short: I would actually prefer if we could store the user's language on sign ups, too.

Reply:
In this case I would remove the not-NULL constraint and use the browser locale or the default locale when the other options are all nil.

Reply:
Alright, then we use the automatically-derived language as long as the use hasn't made a choice themselves. Fine for me :+1:.